nhrpd/nhrp_nhs.c

```diff
@@ -197,7 +197,9 @@ static int nhrp_reg_send_req(struct thread *t)
 
 	/* FIXME: push CIE for each local protocol address */
 	cie = nhrp_cie_push(zb, NHRP_CODE_SUCCESS, NULL, NULL);
-	cie->prefix_length = 0xff;
+	/* As per RFC2322, the prefix length is the widest acceptable destination protocol
+	 * address prefix. However, if "Uniqueness" bit is set then it must be 255 */
+	cie->prefix_length = (if_ad->flags & NHRP_IFF_REG_NO_UNIQUE) ? 8 * sockunion_get_addrlen(dst_proto) : 0xff;
```

Comment:
could you please give more explanation on the RFC.
uniqueness does not necessarily means /32 or /128 mask.

Reply:
RFC quoted there should be RFC 2332 not 2322. The uniqueness bit is described here: https://tools.ietf.org/html/rfc2332#section-5.2.1

The relevant text is "Note that when this bit is set in an NHRP Registration Request, only a single CIE may be specified in the NHRP Registration Request and that CIE must have the Prefix Length field set to 0xFF."

This is pretty unambiguous, and clearly FRR nhrpd was previously not following the RFC.

Without this change we were facing interoperability issues with cisco hub, where the cisco hub added an nhrp cache entry for with end point address of 0.0.0.0, and frr ended up endlessly retrying registration.

Reply:
does that mean that you don't set ipv4 address / 32 on your local tunnel interface ?

Reply:
so , the prefix len will be set to 0 in case no unique flag is set.

Reply:
I think a different section of the RFC explains this better:

Prefix Length
       If a CIE is specified in the NHRP Resolution Request then the
       Prefix Length field may be used to qualify the widest acceptable
       prefix which may be used to satisfy the NHRP Resolution Request.
       In the case of NHRP Resolution Request/Reply, the Prefix Length
       specifies the equivalence class of addresses which match the
       first "Prefix Length" bit positions of the Destination Protocol
       Address.  If the "U" bit is set in the common header then this
       field MUST be set to 0xFF.

So when unique bit is not set, we should be sending 32 in this field not 255. When 255 is there, cisco devices get confused and add an invalid nhrp entry:

*Oct 16 05:38:07.347: NHRP: Receive Registration Request via Tunnel0 vrf global(0x0), packet size: 92
*Oct 16 05:38:07.347:  (F) afn: AF_IP(1), type: IP(800), hop: 1, ver: 1
*Oct 16 05:38:07.347:      shtl: 4(NSAP), sstl: 0(NSAP)
*Oct 16 05:38:07.347:      pktsz: 92 extoff: 52
*Oct 16 05:38:07.347:  (M) flags: "nat ", reqid: 1
*Oct 16 05:38:07.347:      src NBMA: 10.5.0.126
*Oct 16 05:38:07.347:      src protocol: 10.0.0.1, dst protocol: 10.0.0.1
*Oct 16 05:38:07.347:  (C-1) code: no error(0)
**Oct 16 05:38:07.347:        *prefix: 255, mtu: 0, hd_time: 7200**
*Oct 16 05:38:07.347:        addr_len: 0(NSAP), subaddr_len: 0(NSAP), proto_len: 0, pref: 0
*Oct 16 05:38:07.347: NHRP-DETAIL: netid_in = 1, to_us = 1
*Oct 16 05:38:07.347: NHRP: Tunnels gave us pak src: 10.5.0.126
*Oct 16 05:38:07.347: NHRP: Could not find AVL node for vrf:global(0x0)
*Oct 16 05:38:07.347: NHRP: Could not find AVL node for vrf:global(0x0)
*Oct 16 05:38:07.347: NHRP-DETAIL: Cache init: NHRP AVL root is NULL
*Oct 16 05:38:07.347: NHRP: Inserted node for vrf global(0x0)
*Oct 16 05:38:07.347: NHRP: Initialized cache radix head for vrf global(0x0)
*Oct 16 05:38:07.347: NHRP-RT: Attempting to create instance PDB for vrf global(0x0)(0x0)
***Oct 16 05:38:07.347: NHRP-CACHE: Tunnel0: Cache add for target 0.0.0.0/0 vrf global(0x0) label none next-hop 0.0.0.0**
***Oct 16 05:38:07.347:            10.5.0.126**
***Oct 16 05:38:07.347: NHRP: Adding Tunnel Endpoints (VPN: 0.0.0.0, NBMA: 10.5.0.126)**
*Oct 16 05:38:07.351: NHRP: Successfully attached NHRP subblock for Tunnel Endpoints (VPN: 0.0.0.0, NBMA: 10.5.0.126)

With this change in place, the cisco device correctly adds the entry:

...
Oct 16 05:46:14.027:        *prefix: 32, mtu: 17912, hd_time: 7200
...
*Oct 16 05:46:14.027: NHRP-CACHE: Tunnel0: Cache add for target 10.0.0.1/32 vrf global(0x0) label none next-hop 10.0.0.113
*Oct 16 05:46:14.027:            10.5.0.126
*Oct 16 05:46:14.027: NHRP: Adding Tunnel Endpoints (VPN: 10.0.0.1, NBMA: 10.5.0.126)